File: Py_prog/merge_sort.py
def merge(l1,l2):
        t1,t2=0,0
        (c,m,n)=([],len(l1),len(l2))
        #t1,t2=0,0
        while t1+t2<m+n:
            if t1==m:
                c.append(l2[t2])
                t2+=1
            elif t2==n:
                c.append(l1[t1])
                t1+=1
            elif l1[t1]>l2[t2]:
                c.append(l2[t2])
                t2+=1
            elif l1[t1]<l2[t2]:
                c.append(l1[t1])
                t1+=1
        
        #for sorted lists
        print('#',c,'#')
        #for sorted lists
        
        return c

def mergesort(a,left,right):
    if right-left<=1:
        return a[left:right]
    if right- left>1:
        mid=(left+right)//2

        #for sublists
        print(a[left:mid],a[mid:right])
        #for sublists

        l=mergesort(a,left,mid)
        r=mergesort(a,mid,right)
        
        return merge(l,r)

# m=int(input('enter the length of first list:'))
#n=int(input('enter the length of second list:'))

# l1=[int(input('enter the element-a:')) for x in range(m)]
# l1.sort()
l1 = [76,90,21,45,56,9]

print(l1)
l=mergesort(l1,0,len(l1))
print(l)

File: ALGO'S/merge_sort.c
#include<stdio.h>

void merge(int a[],int l,int m,int h){

        int n1=m-l+1,n2=h-m;
        int l1[m-l+1],l2[h-m];
        for (int i=0;i<n1;i++){
            l1[i]=a[l+i];
        }
        for (int i=0;i<n2;i++){
            l2[i]=a[m+i+1];
        }

        int i=0,j=0,k=l;

        while(i<n1 && j<n2){
            if (l1[i]>l2[j]){
                a[k++]=l2[j++];
            }
            else{
                a[k++]=l1[i++];
            }
        }

        while(i<n1){
            a[k++]=l1[i++];
        }
        while(j<n2){
            a[k++]=l2[j++];
        }

}


void mergesort(int a[],int l,int h)
{
    if (l<h){

        int mid=(l+h)/2;
        mergesort(a,l,mid);
        mergesort(a,mid+1,h);
        merge(a,l,mid,h);
    }
}

void print(int a[],int n){
    for (int i=0;i<n;i++){
        printf("%d ",a[i]);
    }
    printf("\n");
}

int main(){

    int a[]={3,1,2,4,7,6,5};
    print(a,7);
    mergesort(a,0,7);
    print(a,7);

}